Add composable activation checkpointing

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #87664
* #87954

This is a composable activation checkpointing API. Unlike functional
activation checkpointing APIs, this one does not require changing
model source code. Unlike ``nn.Module`` wrapper activation checkpointing
APIs, this one does not modify model structure or fully-qualified names
either. Under the hood, it registers activation checkpointing logic as pre-
and post-forward hooks.

mrshenli committed:
[WIP] Add composable checkpointing

[ghstack-poisoned]

mrshenli committed:
Update on "[WIP] Add composable checkpointing"




[ghstack-poisoned]

zhaojuanmao commented on torch/distributed/_composable/checkpoint.py:
```diff
@@ -0,0 +1,127 @@
+import torch
```
nit: wondering whether we can make the file name to be like 'checkpoint_activation.py"? so that it can be differentiated from model checkpoint in the future

mrshenli replied:
sure, will update.

mrshenli committed:
Update on "[WIP] Add composable checkpointing"




[ghstack-poisoned]

mrshenli committed:
Update on "[WIP] Add composable checkpointing"




[ghstack-poisoned]

yhcharles commented on torch/distributed/_composable/contract.py:
```diff
@@ -0,0 +1,97 @@
+        all_state.setdefault(func, {})
+
+        orig_named_params = OrderedDict(module.named_parameters())
+        orig_named_buffers = OrderedDict(module.named_buffers(remove_duplicate=False))
```
It seems `module.named_buffers()` has no arg of `remove_duplicate`:
https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.named_buffers

yhcharles commented on torch/distributed/_composable/contract.py:
```diff
@@ -0,0 +1,97 @@
+            updated = module
+
+        new_named_params = OrderedDict(updated.named_parameters())
+        new_named_buffers = OrderedDict(updated.named_buffers(remove_duplicate=False))
```
Ditto

yhcharles commented on torch/distributed/_composable/contract.py:
```diff
@@ -0,0 +1,97 @@
+            "Composable distributed API implementations cannot modify "
+            "parameter FQNs. \n"
+            f"Original FQNs: {list(orig_named_params.keys())}, \n"
+            f"Updated FQNs: {list(OrderedDict(module.named_parameters()).keys())}"
```
Do you mean `new_named_params.keys()` here?

yhcharles commented on torch/distributed/_composable/contract.py:
```diff
@@ -0,0 +1,97 @@
+            f"Original FQNs: {list(orig_named_buffers.keys())}, \n"
+            f"Updated FQNs: {list(OrderedDict(module.named_buffers()).keys())}"
+        )
+        assert new_named_modules.keys() == new_named_modules.keys(), (
```
Typo? `new_named_modules` -> `orig_named_modules`?

yhcharles commented on torch/distributed/_composable/contract.py:
```diff
@@ -0,0 +1,97 @@
+            "Composable distributed API implementations cannot modify "
+            "buffer FQNs. \n"
+            f"Original FQNs: {list(orig_named_buffers.keys())}, \n"
+            f"Updated FQNs: {list(OrderedDict(module.named_buffers()).keys())}"
```
Ditto

yhcharles commented on torch/distributed/_composable/contract.py:
```diff
@@ -0,0 +1,97 @@
+            "Composable distributed API implementations cannot modify "
+            "submodule FQNs. \n"
+            f"Original FQNs: {orig_named_modules.keys()}, \n"
+            f"Updated FQNs: {OrderedDict(module.named_modules()).keys()}"
```
Ditto

yhcharles commented on torch/distributed/_composable/checkpoint.py:
```diff
@@ -0,0 +1,127 @@
+    Disable hooks installed by checkpoint to avoid unintentional recursion
+    during backward recomputation.
+    """
+    orig_enable_hook = checkpoint.get_state(module, ENALBE_HOOK_KEY)
```
Do we need to import `checkpoint` to use this name?

mrshenli committed:
Update on "[WIP] Add composable checkpointing"




[ghstack-poisoned]

mrshenli committed:
Update on "[WIP] Add composable checkpointing"




[ghstack-poisoned]

mrshenli committed:
Update on "[WIP] Add composable checkpointing"




[ghstack-poisoned]

mrshenli committed:
Update on "[WIP] Add composable checkpointing"




[ghstack-poisoned]

mrshenli committed:
Update on "[WIP] Add composable checkpointing"




[ghstack-poisoned]

mrshenli committed:
Update on "Add composable activation checkpointing"


This is a composable activation checkpointing API. Unlike functional
activation checkpointing APIs, this one does not require changing
model source code. Unlike ``nn.Module`` wrapper activation checkpointing
APIs, this one does not modify model structure or fully-qualified names
either. Under the hood, it registers activation checkpointing logic as pre-
and post-forward hooks.

[ghstack-poisoned]

mrshenli committed:
Update on "Add composable activation checkpointing"


This is a composable activation checkpointing API. Unlike functional
activation checkpointing APIs, this one does not require changing
model source code. Unlike ``nn.Module`` wrapper activation checkpointing
APIs, this one does not modify model structure or fully-qualified names
either. Under the hood, it registers activation checkpointing logic as pre-
and post-forward hooks.

[ghstack-poisoned]

mrshenli committed:
Update on "Add composable activation checkpointing"


This is a composable activation checkpointing API. Unlike functional
activation checkpointing APIs, this one does not require changing
model source code. Unlike ``nn.Module`` wrapper activation checkpointing
APIs, this one does not modify model structure or fully-qualified names
either. Under the hood, it registers activation checkpointing logic as pre-
and post-forward hooks.

[ghstack-poisoned]

mrshenli committed:
Update on "Add composable activation checkpointing"


This is a composable activation checkpointing API. Unlike functional
activation checkpointing APIs, this one does not require changing
model source code. Unlike ``nn.Module`` wrapper activation checkpointing
APIs, this one does not modify model structure or fully-qualified names
either. Under the hood, it registers activation checkpointing logic as pre-
and post-forward hooks.

[ghstack-poisoned]

zhaojuanmao reviewed (approved):
awesome!

mrshenli commented:
@pytorchbot merge -g

awgu commented:
Failure looks real, but the error may have been swallowed without traceback:
```
======================================================================
ERROR [0.001s]: test_tensor_only_gpu (__main__.TestCheckpoint)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_distributed.py", line 146, in wrapper
    sys.exit(TEST_SKIPS[f"multi-gpu-{x}"].exit_code)
SystemExit: 75
```

mrshenli committed:
Update on "Add composable activation checkpointing"


This is a composable activation checkpointing API. Unlike functional
activation checkpointing APIs, this one does not require changing
model source code. Unlike ``nn.Module`` wrapper activation checkpointing
APIs, this one does not modify model structure or fully-qualified names
either. Under the hood, it registers activation checkpointing logic as pre-
and post-forward hooks.

[ghstack-poisoned]

mrshenli commented:
@pytorchbot merge -g

mrshenli commented:
@pytorchbot merge -g
